fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java

```diff
@@ -0,0 +1,194 @@
+                    .put(FunctionSet.SUM, Pair.create(FunctionSet.SUM, FunctionSet.SUM))
+                    .put(FunctionSet.HLL_UNION, Pair.create(FunctionSet.HLL_UNION, FunctionSet.HLL_UNION))
+                    .put(FunctionSet.NDV, Pair.create(FunctionSet.HLL_UNION, FunctionSet.NDV))
+                    .put(FunctionSet.BITMAP_UNION, Pair.create(FunctionSet.BITMAP_UNION, FunctionSet.BITMAP_UNION))
+                    .put(FunctionSet.BITMAP_UNION_COUNT,
+                            Pair.create(FunctionSet.BITMAP_UNION, FunctionSet.BITMAP_UNION_COUNT))
+                    .put(FunctionSet.PERCENTILE_UNION,
+                            Pair.create(FunctionSet.PERCENTILE_UNION, FunctionSet.PERCENTILE_UNION))
+                    .build();
+
+    public GroupByCountDistinctRewriteRule() {
+        super(RuleType.TF_REWRITE_GROUP_BY_COUNT_DISTINCT,
+                Pattern.create(OperatorType.LOGICAL_AGGR, OperatorType.LOGICAL_OLAP_SCAN));
+    }
+
+    @Override
+    public boolean check(OptExpression input, OptimizerContext context) {
+        LogicalAggregationOperator aggregate = (LogicalAggregationOperator) input.getOp();
+        LogicalOlapScanOperator scan = (LogicalOlapScanOperator) input.getInputs().get(0).getOp();
+
+        Collection<CallOperator> calls = aggregate.getAggregations().values();
+
+        List<CallOperator> distinctList = calls.stream().filter(this::isDistinct).collect(Collectors.toList());
+        List<CallOperator> otherList = calls.stream().filter(c -> !isDistinct(c)).collect(Collectors.toList());
+
+        // check other function
+        if (!otherList.isEmpty() && !otherList.stream().map(c -> c.getFunction().getFunctionName().getFunction())
+                .allMatch(OTHER_FUNCTION_TRANS::containsKey)) {
+            return false;
+        }
+
+        // check distinct function
+        if (distinctList.isEmpty() || !distinctList.stream().map(f -> f.getFunction().getFunctionName().getFunction())
+                .allMatch(DISTINCT_FUNCTION_TRANS::containsKey)) {
+            return false;
+        }
+
+        // check distinct column only one
+        if (distinctList.stream().anyMatch(f -> f.getChildren().size() > 1)) {
+            return false;
+        }
+
+        if (!distinctList.stream().map(f -> f.getChild(0)).allMatch(ScalarOperator::isColumnRef)) {
+            return false;
+        }
+
+        List<ColumnRefOperator> distinctColumns = distinctList.stream().map(f -> f.getChild(0))
+                .map(c -> (ColumnRefOperator) c).distinct().collect(Collectors.toList());
+        if (distinctColumns.size() != 1) {
+            return false;
+        }
+
+        // check distribution satisfy scan node
+        List<Integer> groupBy = aggregate.getGroupingKeys().stream().map(ColumnRefOperator::getId)
+                .collect(Collectors.toList());
+
+        if (groupBy.isEmpty() || groupBy.containsAll(scan.getDistributionSpec().getShuffleColumns())) {
```

Comment:
Why return false when groupBy columns satisfied scan's shuffle columns?

Reply:
don't need rewrite by this rule, origin plan may better